src/infrastructure/acknowledgment.rs

```diff
@@ -0,0 +1,249 @@
+use crate::packet::OrderingGuarantee;
+use crate::packet::SequenceNumber;
+use crate::sequence_buffer::SequenceBuffer;
+use std::collections::HashMap;
+
+const REDUNDANT_PACKET_ACKS_SIZE: u16 = 32;
+
+/// Responsible for handling the acknowledgment of packets.
+pub struct AcknowledgmentHandler {
+    // Local sequence number which we'll bump each time we send a new packet over the network
+    sequence_number: SequenceNumber,
+    // The last acked sequence number of the packets we've sent to the remote host.
+    remote_ack_sequence_num: SequenceNumber,
+    // Using a Hashmap to track every packet we send out so we can ensure that we can resend when
+    // dropped.
+    sent_packets: HashMap<u16, SentPacket>,
+    // However, we can only reasonably ack up to REDUNDANT_PACKET_ACKS_SIZE + 1 packets on each
+    // message we send so this should be REDUNDANT_PACKET_ACKS_SIZE large.
+    received_packets: SequenceBuffer<ReceivedPacket>,
+}
+
+impl AcknowledgmentHandler {
+    /// Constructs a new `AcknowledgmentHandler` with which you can perform acknowledgment operations.
+    pub fn new() -> Self {
+        AcknowledgmentHandler {
+            sequence_number: 0,
+            remote_ack_sequence_num: 0,
+            sent_packets: HashMap::new(),
+            received_packets: SequenceBuffer::with_capacity(REDUNDANT_PACKET_ACKS_SIZE),
+        }
+    }
+
+    /// Returns the next sequence number to send.
+    pub fn local_sequence_num(&self) -> SequenceNumber {
+        self.sequence_number
+    }
+
+    /// Returns the last sequence number received from the remote host (+1)
+    pub fn remote_sequence_num(&self) -> SequenceNumber {
+        self.received_packets.sequence_num()
+    }
+
+    /// Returns the ack_bitfield corresponding to which of the past 32 packets we've
+    /// successfully received.
+    pub fn ack_bitfield(&self) -> u32 {
+        let most_recent_remote_seq_num: u16 = self.remote_sequence_num().wrapping_sub(1);
```

Comment:
`last_remote_seq_num`?

Reply:
I wanted to make this more specific in that it should be the sequence number that is furthest ahead rather than the last one we've received.